Posix: fs: Add `rmdir` and `remove` posix API support

karthi012 opened the pull request:
Add POSIX API support for rmdir and remove and along with test suites.

karthi012 committed:
lib: posix: fs: add rmdir support

Implementation of rmdir, Posix style file
system API to remove directory

Signed-off-by: Karthikeyan Krishnasamy <karthikeyan@linumiz.com>

karthi012 committed:
tests: posix: fs: add test suite for rmdir

add test suite for POSIX API rmdir

Signed-off-by: Karthikeyan Krishnasamy <karthikeyan@linumiz.com>

karthi012 committed:
lib: libc: stdio: add support remove api

Implement remove api to zephyr's native c library

Signed-off-by: Karthikeyan Krishnasamy <karthikeyan@linumiz.com>

de-nordic commented on tests/posix/fs/src/test_fs_file.c:
```diff
@@ -287,6 +300,17 @@ ZTEST(posix_fs_file_test, test_fs_unlink)
+ *
+ * @details Test deletes a file through POSIX remove API.
+ */
+ZTEST(posix_fs_file_test, test_fs_remove)
+{
+	zassert_true(test_file_open() == TC_PASS);
+	zassert_true(test_file_remove() == TC_PASS);
+}
+
```
/* FIXME: restructure tests as per #46897 */ ; can we start here by not doing that and directly use
`zassert_equal()`, `zassert_true()` here wit assert message?
Also, shouldn't negative tests be also here, like incorrect path provided or non existing mount point path?

cfriedt replied:
Yeah, I agree - let's let the dinosaurs die (i.e. start writing good tests and have less to refactor later).

Definitely want to validate behaviour in the case of negative / degenerate input as well.

cfriedt commented on tests/posix/fs/src/test_fs_dir.c:
```diff
@@ -88,6 +88,30 @@ static int test_lsdir(const char *path)
 	return res;
 }
 
+static int test_rmdir(void)
+{
+	int res;
+
+	TC_PRINT("\nmkdir tests:\n");
+
+	res = mkdir(TEST_DIR, S_IRWXG);
+	if (res) {
+		TC_PRINT("Error creating dir[%d]\n", res);
+		return res;
+	}
+
+	TC_PRINT("\nrmdir tests:\n");
+
+	res = rmdir(TEST_DIR);
+	if (res) {
+		TC_PRINT("Error removing dir[%d]\n", res);
+		return res;
+	}
+
+	TC_PRINT("\nremoved dir: %s\n",  TEST_DIR);
+	return res;
+}
```
@karthi012 - this sort of pattern should suffice for the more modern style of tests. Can you please make some adjustments this this test and similar ones to the `remove` test?

```suggestion
ZTEST(posix_fs_dir_test, test_rmdir)
{
	/* degenerate cases */
	// rmdir ops that are expected to fail in various predictable ways 

	/* happy path */
	zassert_ok(mkdir(TEST_DIR, S_IRWXG), "Error creating dir: %d", errno);
	zassert_ok(rmdir(TEST_DIR), "Error removing dir: %d\n", errno);
}
```

cfriedt commented on lib/posix/options/fs.c:
```diff
@@ -438,3 +438,23 @@ int ftruncate(int fd, off_t length)
+ *
+ * See IEEE 1003.1
+ */
+int remove(const char *path)
```
This one is a bit tricky because remove has been a part of ISO C since 1989.

That means, this function should probably go into the common libc.

https://en.cppreference.com/w/c/io/remove

And in order to eliminate dependency cycles at the API level, the libc version of this should call the Zephyr `fs_unlink()` function directly rather than the POSIX `unlink()`.

If `CONFIG_FS` is not set, then `remove()` can return -1 and set errno to `ENOTSUP`.

With that, `rmdir()` and `remove()` still satisfy all of the requirements from POSIX (i.e. that they are effectively both calls to the same `fs_unlink()`.

Thoughts, @de-nordic ?

de-nordic replied:
I agree with @cfriedt  that we should rather directly call to system specific function, so ‘fs_unlink’ should be called.

As far as I understand https://pubs.opengroup.org/onlinepubs/9699919799/functions/remove.html#tag_16_485 it is part of both, I don’t mind having it implemented as part of POSIX; we can discuss further the right place, but at least we will have it, till we agree.

Shouldn’t we return ‘-ENOSYS’ https://docs.zephyrproject.org/latest/develop/api/design_guidelines.html#return-codes.

karthi012 replied:
@cfriedt  I can remove the `remove` api as it implement in the libc.

cfriedt replied:
It isn't implemented (yet) in the common libc, so I would suggest moving it there.

karthi012 replied:
@cfriedt  Are you Pointing to the zephyr native libc or something.  I think `remove` is implementated in the picolibc`modules/lib/picolibc/./newlib/libc/stdio/remove.c`.

cfriedt replied:
> @cfriedt  Are you Pointing to the zephyr native libc or something.

In `lib/libc/common`

> I think `remove` is implementated in the picolibc`modules/lib/picolibc/newlib/libc/stdio/remove.c`.

```cpp
int
remove (
       const char *filename)
{
  if (unlink (filename) == -1)
    return -1;

  return 0;
}
```

And what happens when `unlink()` is unavailable?

This is one of the reasons why we have `lib/libc/common` in Zephyr.

Since `remove()` is part of ISO C, and since POSIX is not required for ISO C applications, we need to implement `remove()` via Zephyr's native APIs (i.e. `fs_unlink()`). With that both POSIX and C are share a mutual dependency instead of a dependency cycle.

cfriedt commented on tests/posix/fs/src/test_fs_file.c:
```diff
@@ -287,6 +287,18 @@ ZTEST(posix_fs_file_test, test_fs_unlink)
+ *
+ * @details Test deletes a file through POSIX remove API.
+ */
+ZTEST(posix_fs_file_test, test_fs_remove)
```
This test should be moved to the common libc test suite.

karthi012 commented on tests/posix/fs/src/test_fs_dir.c:
```diff
@@ -126,3 +125,26 @@ ZTEST(posix_fs_dir_test, test_fs_readdir)
+	zassert_ok(mkdir(TEST_DIR, IRWXG), "Error creating dir: %d", errno);
+	zassert_not_equal(open(TEST_DIR_FILE, O_CREAT | O_RDWR), -1,
+			  "Error creating file: %d", errno);
+	zassert_not_ok(rmdir(TEST_DIR), "Error Non empty dir removed");
```
@cfriedt Zephyr FAT fs does not report proper errno matched with the POSIX return values. But littlefs is return.
```static int translate_error(int error) 
{ 
        switch (error) { 
        case FR_OK: 
                return 0; 
        case FR_NO_FILE: 
        case FR_NO_PATH: 
        case FR_INVALID_NAME: 
                return -ENOENT; 
        case FR_DENIED: 
                return -EACCES; 
        case FR_EXIST: 
                return -EEXIST; 
        case FR_INVALID_OBJECT: 
                return -EBADF; 
        case FR_WRITE_PROTECTED: 
                return -EROFS; 
        case FR_INVALID_DRIVE: 
        case FR_NOT_ENABLED: 
        case FR_NO_FILESYSTEM: 
                return -ENODEV; 
        case FR_NOT_ENOUGH_CORE: 
                return -ENOMEM; 
        case FR_TOO_MANY_OPEN_FILES: 
                return -EMFILE; 
        case FR_INVALID_PARAMETER: 
                return -EINVAL; 
        case FR_LOCKED: 
        case FR_TIMEOUT: 
        case FR_MKFS_ABORTED: 
        case FR_DISK_ERR: 
        case FR_INT_ERR: 
        case FR_NOT_READY: 
                return -EIO; 
        } 
 
        return -EIO; 
}```

cfriedt replied:
As long as we use the fs API from zephyr, POSIX should expect consistent return values and error numbers, regardless of the underlying filesystem. Understandably, FAT is missing some features, so if e.g. the underlying FS is FAT and we look for some kind of permission value, I would guess we get `-ENOTSUP` back.

cfriedt replied:
If there are not already consistent return values from the fs subsystem, then we can likely consider this a future problem.

cfriedt commented on tests/posix/fs/testcase.yaml:
```diff
@@ -3,6 +3,7 @@ common:
   arch_exclude:
     - nios2
   platform_exclude:
+    - esp32s3_touch_lcd_1_28/esp32s3/appcpu
```
What was the reason for this exclusion?

karthi012 replied:
@cfriedt  RAM overflowed in this target.

cfriedt replied:
Would be best to add a line like this:

```yaml
min_ram: 16
```

And then we would not need to filter by platform, so the specific exclusion can be removed.

See https://docs.zephyrproject.org/latest/develop/test/twister.html

cfriedt commented on lib/libc/common/Kconfig:
```diff
@@ -82,3 +82,9 @@ config COMMON_LIBC_THRD
+
+config COMMON_LIBC_REMOVE
+	bool
+	default y if !POSIX_FS
```
I think this should just be `default y`.

cfriedt commented on lib/libc/common/source/stdio/remove.c:
```diff
@@ -0,0 +1,30 @@
+int remove(const char *path)
+{
+	int ret;
+
```
Something like this should work in case the FS is not enabled.

```suggestion
if (!IS_ENABLED(CONFIG_FS)) {
  errno = ENOTSUP;
  return -1;
}
```

de-nordic replied:
Yeah, but do the other part of remove in else. Otherwise you just carry the rest of function that will never be executed or start triggering static analysis that there is code after `if (1) { }`

cfriedt replied:
Dead code should be stripped by the compiler, so there should be no need for an explicit `else`

cfriedt commented on doc/services/portability/posix/option_groups/index.rst:
```diff
@@ -341,7 +341,7 @@ POSIX_FILE_SYSTEM
     rename(), yes
     rewinddir(),
-    rmdir(),
+    rmdir(), yes
```
it would make sense to mark `remove()` as supported as well.

cfriedt commented on tests/posix/fs/src/test_fs_dir.c:
```diff
@@ -28,7 +28,6 @@ static int test_mkdir(void)
 	}
 
 	res = open(TEST_DIR_FILE, O_CREAT | O_RDWR);
-
```
@karthi012 it looks like the blank line needs to be added back to this position.

```shell
git add x
git commit --fixup b2bd082
git rebase -i --autosquash b2bd082~1
```

karthi012 replied:
@cfriedt  Apologies.

cfriedt commented:
Would be good to run tests through `twister` as well prior to pushing again.

Something like this should work
```shell
twister -i -T tests/posix -T tests/lib/c_lib
```

karthi012 commented:
> Would be good to run tests through `twister` as well prior to pushing again.
> 
> Something like this should work
> 
> ```shell
> twister -i -T tests/posix -T tests/lib/c_lib
> ```

Thanks for the twister test commands.

karthi012 committed:
tests: lib: c_lib: add test suite for remove api

Add test suite for remove of zephyr native libc

Signed-off-by: Chris Friedt <cfriedt@tenstorrent.com>
Signed-off-by: Karthikeyan Krishnasamy <karthikeyan@linumiz.com>

cfriedt commented on lib/libc/common/Kconfig:
```diff
@@ -82,3 +82,9 @@ config COMMON_LIBC_THRD
+
+config COMMON_LIBC_REMOVE
+	bool
+	default y
```
As [previously mentioned](https://github.com/zephyrproject-rtos/zephyr/pull/71923#discussion_r1590005588), both picolibc and newlib call `unlink()` in their implementations, which relies on POSIX.

AFAIK, picolibc only supports `unlink()` via semihosting?

So effectively `remove()` is not provided by any libc.

Perhaps a better value for this is
```
depends on !SEMIHOSTING
default y if FS
```

cfriedt commented on tests/lib/c_lib/stdio/testcase.yaml:
```diff
@@ -0,0 +1,8 @@
+common:
+  tags:
+    - clib
+  min_ram: 128
+tests:
+  libraries.libc.stdio: {}
+  libraries.libc.stdio.remove:
+    filter: not CONFIG_NATIVE_LIBC
```
This should help us filter-out native_posix in CI for all build configurations, and will also test against multiple libc's.


```suggestion
common:
  tags:
    - clib
  min_ram: 128
  filter: not CONFIG_NATIVE_LIBC
tests:
  libraries.libc.common.stdio.minimal:
    tags: minimal_libc
    filter: CONFIG_MINIMAL_LIBC_SUPPORTED
    extra_configs:
      - CONFIG_MINIMAL_LIBC=y
  libraries.libc.common.stdio.picolibc:
    tags: picolibc
    filter: CONFIG_PICOLIBC_SUPPORTED
    extra_configs:
      - CONFIG_PICOLIBC=y
  libraries.libc.common.stdio.newlib:
    tags: newlib
    filter: TOOLCHAIN_HAS_NEWLIB == 1
    extra_configs:
      - CONFIG_NEWLIB_LIBC=y
```

cfriedt reviewed (commented):
It's pretty close. Just some yaml changes to avoid the CI failure.

cfriedt commented:
@karthi012 - please be sure to rebase on latest `main`

karthi012 commented:
> @karthi012 - please be sure to rebase on latest `main`

Apologies. I have rebased branch with current mainline when i pushed the changes.

cfriedt reviewed (commented):
I think there was one coding style flag in CI. A false positive about reserved identifiers. Otherwise, LGTM

karthi012 commented:
@cfriedt Changes have been pushed. Thanks for your help.

cfriedt reviewed (dismissed):
Looks good to me - there is a false positive CI warning about using the reserved name which this PR is implementing.

cfriedt commented:
@karthi012 - please rebase on latest main - it seems that some deprecated DT binding is causing failures here.

cfriedt commented:
@dleach02 @henrikbrixandersen - any ideas about this one? It was green until recently and @karthi012 recently rebased on main, IIUC.

```
-- Found BOARD.dts: /__w/zephyr/zephyr/boards/nxp/imx93_evk/imx93_evk_mimx9352_a55.dts
-- Found devicetree overlay: /__w/zephyr/zephyr/tests/lib/c_lib/stdio/app.overlay
devicetree error: 'bus-speed' is marked as deprecated in 'properties:' in /__w/zephyr/zephyr/dts/bindings/can/nxp,flexcan-fd.yaml for node /can@425b0000.
```

ycsin reviewed (dismissed):
Compliance error should be a false positive

aescolar commented:
@karthi012 rebase needed

karthi012 committed:
doc: posix: mark rmdir and remove as supported

Mark POSIX API rmdir and remove as supported in
POSIX options group Documentation

Signed-off-by: Karthikeyan Krishnasamy <karthikeyan@linumiz.com>

cfriedt commented:
@karthi012 - I'm sorry that your PR has grown stale multiple times and has been excluded e.g. from the last release. I think this is mainly due to false positives in Zephyr's CI.

As a workaround, feel free to include [this commit](https://github.com/zephyrproject-rtos/zephyr/pull/77104/commits/b4dfbdbb4d5e5442c60e773a7478326a46815946) in your PR.
